File: leetcode/01022_sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        return helper(root, 0);
    }

private:
    int helper(TreeNode* node, int accumulate) {
        if (node == NULL) {
            return 0;
        }

        accumulate = accumulate * 2 + node->val;
        if (!node->left && !node->right) {
            return accumulate;
        }

        return helper(node->left, accumulate) + helper(node->right, accumulate);
    }
};

File: leetcode/01022_sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        ret = 0

        def traverse(node, accumulate):
            nonlocal ret
            accumulate = accumulate * 2 + node.val
            if not (node.left or node.right):
                ret += accumulate
            else:
                if node.left:
                    traverse(node.left, accumulate)
                if node.right:
                    traverse(node.right, accumulate)

        traverse(root, 0)
        return ret
